extmod/uasyncio/taskgroup.py

```diff
@@ -0,0 +1,244 @@
+                #    async with TaskGroup() as g:
+                #        g.create_task(...)
+                #        await ...  # <- CancelledError
+                #
+                # or there's an exception in "async with":
+                #
+                #    async with TaskGroup() as g:
+                #        g.create_task(...)
+                #        1 / 0
+                #
+                self._abort()
+
+        # We use while-loop here because "self._on_completed"
+        # can be cancelled multiple times if our parent task
+        # is being cancelled repeatedly (or even once, when
+        # our own cancellation is already in progress)
+        while self._tasks:
+            if self._on_completed is None:
+                self._on_completed = event.Event()
+
+            try:
+                await self._on_completed.wait()
+            except core.CancelledError as ex:
+                if not self._aborting:
+                    # Our parent task is being cancelled:
+                    #
+                    #    async def wrapper():
+                    #        async with TaskGroup() as g:
+                    #            g.create_task(foo)
+                    #
+                    # "wrapper" is being cancelled while "foo" is
+                    # still running.
+                    propagate_cancellation_error = ex
+                    self._abort()
+
+            self._on_completed = None
+
+        assert not self._tasks
+
+        if self._base_error is not None:
+            raise self._base_error
+
+        if propagate_cancellation_error is not None:
+            # The wrapping task was cancelled; since we're done with
+            # closing all child tasks, just propagate the cancellation
+            # request now.
+            raise propagate_cancellation_error
+
+        if et is not None and et is not core.CancelledError:
+            self._errors.append(exc)
+
+        if self._errors:
+            # Exceptions are heavy objects that can have object
+            # cycles (bad for GC); let's not keep a reference to
+            # a bunch of them.
+            errors = self._errors
+            self._errors = None
+
+            if len(errors) == 1:
+                me = errors[0]
+            else:
+                EGroup = core.ExceptionGroup
+                for err in errors:
+                    if not isinstance(err, Exception):
+                        EGroup = core.baseExceptionGroup
+                        break
+                me = EGroup("unhandled errors in a TaskGroup", errors)
+            raise me
+
+    def create_task(self, coro):
+        if not self._entered:
+            raise RuntimeError(f"TaskGroup {repr(self)} has not been entered")
+        if self._exiting and not self._tasks:
+            raise RuntimeError(f"TaskGroup {repr(self)} is finished")
+
+        k = [None]
+        t = self._loop.create_task(self._run_task(k, coro))
+        k[0] = t  # sigh
```

Comment:
Why does it need to use a list with one element?

Reply:
Because you need to pass the newly created task to `_run_task`, which you can't otherwise do without some uasyncio surgery, which I wanted to avoid.